geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientUserAuths.java

```diff
@@ -52,17 +62,32 @@ public Long putUserAuth(UserAuthAttributes userAuthAttr) {
     return newId;
   }
 
-  public Long putSubject(Subject subject, long existingUniqueId) {
-    final Long newId;
+
+  public long putSubject(@NotNull Subject subject, long existingUniqueId) {
+    final long newId;
     if (existingUniqueId == 0 || existingUniqueId == NOT_A_USER_ID) {
       newId = getNextID();
     } else {
       newId = existingUniqueId;
     }
 
-    Subject oldSubject = uniqueIdVsSubject.put(newId, subject);
-    removeSubject(oldSubject);
-    logger.debug("Subject of {} added.", newId);
+    // we are saving all the subjects that's related to this uniqueId
+    // we cannot immediately log out the old subject of this userId because
+    // it might already be bound to another thread and doing operations. If
+    // we log out that subject immediately, that thread "authorize" would get null principal.
+    synchronized (this) {
```

Comment:
we are preventing user doing `putSubject` `getSubject`, `removeSubject` at the same time. this `synchronize` block is necessary.

Reply:
`CopyOnWriteArrayList` offers built-in synchronization for all write access and it makes a copy of the underlying array for all write access. If you are trying to synchronize read and write access of the list with your own synchronization, it is not necessary to use `CopyOnWriteArrayList`. Because this introduces extra overhead of synchronization and copying the array.

Reply:
that's true, it might be redundant using that list. I can change it to use regular list